x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/autoscaling/AutoscalerService.java

```diff
@@ -0,0 +1,218 @@
+        ClientHelper.executeAsyncWithOrigin(
+            client,
+            ClientHelper.ML_ORIGIN,
+            GetDeploymentStatsAction.INSTANCE,
+            new GetDeploymentStatsAction.Request(deploymentIds),
+            processDeploymentStats
+        );
+    }
+
+    private void processDeploymentStats(GetDeploymentStatsAction.Response statsResponse) {
+        Map<String, Stats> recentStatsByDeployment = new HashMap<>();
+        Map<String, Integer> numberOfAllocations = new HashMap<>();
+
+        for (AssignmentStats assignmentStats : statsResponse.getStats().results()) {
+            numberOfAllocations.put(assignmentStats.getDeploymentId(), assignmentStats.getNumberOfAllocations());
+            for (AssignmentStats.NodeStats nodeStats : assignmentStats.getNodeStats()) {
+                String statsId = assignmentStats.getDeploymentId() + "@" + nodeStats.getNode().getId();
+                Stats lastStats = lastInferenceStatsByDeploymentNode.get(statsId);
+                Stats nextStats = new Stats(
+                    nodeStats.getInferenceCount().orElse(0L),
+                    nodeStats.getPendingCount() == null ? 0 : nodeStats.getPendingCount(),
+                    nodeStats.getErrorCount() + nodeStats.getTimeoutCount() + nodeStats.getRejectedExecutionCount(),
+                    nodeStats.getAvgInferenceTime().orElse(0.0) / 1000.0
+                );
+                lastInferenceStatsByDeploymentNode.put(statsId, nextStats);
+
+                Stats recentStats = (lastStats == null ? nextStats : nextStats.sub(lastStats));
+                recentStatsByDeployment.compute(
+                    assignmentStats.getDeploymentId(),
+                    (key, value) -> value == null ? recentStats : value.add(recentStats)
+                );
+            }
+        }
+
+        for (Map.Entry<String, Stats> deploymentAndStats : recentStatsByDeployment.entrySet()) {
+            String deploymentId = deploymentAndStats.getKey();
+            Stats stats = deploymentAndStats.getValue();
+            Autoscaler autoscaler = autoscalers.get(deploymentId);
+            autoscaler.process(stats, TIME_INTERVAL_SECONDS, numberOfAllocations.get(deploymentId));
+            Integer newNumberOfAllocations = autoscaler.autoscale();
+            if (newNumberOfAllocations != null) {
+                UpdateTrainedModelDeploymentAction.Request updateRequest = new UpdateTrainedModelDeploymentAction.Request(deploymentId);
+                updateRequest.setNumberOfAllocations(newNumberOfAllocations);
+                ClientHelper.executeAsyncWithOrigin(
+                    client,
+                    ClientHelper.ML_ORIGIN,
+                    UpdateTrainedModelDeploymentAction.INSTANCE,
+                    updateRequest,
+                    ActionListener.wrap(
+                        updateResponse -> logger.info(
+                            "Autoscaled deployment [{}] to [{}] allocations.",
```

Comment:
ML has various `Auditor` classes that index important messages in ES, these messages then show up in the ml UI. It's the best method we have to expressing these kinds of internal changes and making them visible to users. See the usage of `SystemAuditor` in `TrainedModelAssignmentClusterService`

Reply:
added SystemAuditor